specs: permissions (#465)
bluesky-social/atproto-website: src/app/[locale]/specs/permission/en.mdx

export const metadata = {
  title: 'Permissions',
  description:
    'Auth Permissions for Account Resources.',
}

# Permissions

> Authorization to atproto resources is described and granted using "permissions". A group of a permissions related to a specific Lexicon namespace (record types and API endpoints) can be bundled together as a "permission set". Both are used in the context of [OAuth](/specs/oauth) to grant client software access to account resources on a PDS. For example, the ability to write records of specific types to the user's public repository, or make authenticated API requests to remote services. Developers declare the permissions their app requires to function, and end users are shown the permissions when granting access to the app.

Author: I went with "Control over atproto resources"

Each resource type has a defined set of parameters that can attenuate the permission. A permission can be represented in a string format (for direct use as an OAuth scope), or as a JSON object (for use in permission sets). Permission sets are published as public lexicon schemas, and are limited to lexicon resources under the same NSID hierarchy.

## Resource Types

Permissions relate to account resources on PDS instances:

- `repo`: Public Repository (records and collections)
- `rpc`: Service Authentication (API calls to external services)
- `blob`: uploaded media files
- `identity`: DID and handle
- `account`: hosting status, email address

There is an additional "meta" resource, `include`, which is used to reference a permission set (see Permission Sets section below).

This section describes these resources in more depth, including all the parameters that can constrain the permission. For each resource, one parameter may be marked as "positional", which impacts the string representation syntax (described in a later section).

> Resources and permissions are described in the abstract, not tied to specific PDS API endpoints. For example, `rpc` permissions describe the ability to make remote service API calls using inter-service authentication. This includes both proxied PDS requests (using HTTP headers) and fetching service auth tokens using `com.atproto.server.getServiceAuth`.

Reviewer: ```suggestion

Resources and permissions are described in the abstract, not tied to specific PDS API endpoints. For example, `rpc` permissions describe the ability to make remote service API calls using inter-service authentication. This includes both proxied PDS requests (using the `atproto-proxy` HTTP header) and fetching service auth tokens using `com.atproto.server.getServiceAuth`.
```

### `repo`

Write access to records in the account's public repository. Can be limited to specific record types (collections) or actions (eg update vs delete).

Parameters:

> - `collection` (string array, required, positional): NSID of record types. Wildcard (`*`) is allowed and gives access to all records. Partial wildcards are not supported (eg, can not use `com.example.*`)

Reviewer: ```suggestion
- `collection` (string array, required, positional): NSID of record types. Wildcard (`*`) is allowed (when the permission is encoded as an oauth scope parameter i.e. `repo:*`) and gives access to all records. Partial wildcards are not supported (eg, can not use `com.example.*`)
```

Reviewer 2: We should what's meant by the parenthetical more explicitly^^

- `action` (string array, optional): defines the set of record operations allowed. Allowed values are `create`, `update`, `delete`. If not defined, all operations are allowed

Examples:

```
# full permission (create, update, delete) on the Bluesky profile record type
repo:app.example.profile

{
  "type": "permission",
  "resource": "repo",
  "collection": ["app.example.profile"]
}

# same as above, with actions being explicit
repo:app.example.profile?action=create&action=update&action=delete

{
  "type": "permission",
  "resource": "repo",
  "collection": ["app.example.profile"],
  "action": ["create", "update", "delete"]
}

# full permissions on all record types
repo:*

{
  "type": "permission",
  "resource": "repo",
  "collection": ["*"]
}

# delete only permission on all record types

> repo:*?action=delete

Reviewer 2: should we add the JSON representation of this so that it matches the other examples?

```

### `rpc`

The ability to make authenticated API requests to remote services. This includes both requesting service auth tokens (JWTs) using the `com.atproto.server.getServiceAuth` endpoint on the PDS, and requests proxied via the PDS.

> The permission is parameterized by the remote endpoint (`lxm`, short for "Lexicon Method") and the identity of the remote service (the audience, `aud`). Permissions must be restricted by one or the other: the ability to call any API endpoint on any service can not be declared with a single permission.

Reviewer 2: ```suggestion
The permission is parameterized by the remote endpoint (`lxm`, short for "Lexicon Method") and the identity of the remote service (the audience, `aud`). Permissions must be restricted by at least one of these parameters. In other words, the ability to call any API endpoint on any service can not be declared with a single permission.
```

When included in a permission set, the audience parameter may be inherited from the overall set.

Parameters:

- `lxm` (string array, required, positional): NSID of API endpoints. Wildcard (`*`) is allowed and gives access to all endpoints. Partial wildcards are not supported (eg, can not use `com.example.*`).
- `aud` (string, semi-required): audience of API requests, as a DID service reference: DID followed by required service type fragment, eg `did:web:api.example.com#srvtype`). Supports wildcard (`*`). Required in granular string representation, and required in lexicon JSON unless `inheritAud` is true.

> - `inheritAud` (boolean, optional): if true, `aud` does not need to be defined.

Reviewer 2: maybe move the note from about down to here & note that this is only relevant in the context of a permission set

Examples:

```
# Submit moderation reports to any service
rpc:app.example.moderation.createReport?aud=*

{
  "type": "permission",
  "resource": "rpc",
  "lxm": ["app.example.moderation.createReport"],
  "aud": ["*"]
}

# Call any RPC method on a specific service
rpc?lxm=*&aud=did:web:api.example.com%23svc_appview

{
  "type": "permission",
  "resource": "rpc",
  "lxm": ["*"],
  "aud": ["did:web:api.example.com#svc_appview"]
}
```

## `blob`

Ability to upload media files (blobs) to PDS.

Parameters:

> - `accept` (array of strings, required): MIME types or partial MIME type glob patterns (`*/*` or `text/*` for example). Same syntax as the `accept` field in the `blob` lexicon type.

Reviewer: note: Should we add a `collections` field here to restrict which collection can be referencing new blobs ?

Example use case would be "allow app X to upload images for your profile; but not for posts"

Reviewer 2: this isn't implemented anywhere right? I think we hold off

Author: I agree we shouldn't add functionality right now, let's just keep blobs broad.

> - `accept` (array of strings, required): MIME types or partial MIME type glob patterns (`*/*` or `text/*` for example). Same syntax as the `accept` field in the `blob` lexicon type.

Reviewer 2: ```suggestion
- `accept` (array of strings, required, positional): MIME types or partial MIME type glob patterns (`*/*` or `text/*` for example). Same syntax as the `accept` field in the `blob` lexicon type.
```

Examples:

```
# Upload any type of blob
blob:*/*

{
  "type": "permission",
  "resource": "blob",
  "accept": ["*/*"]
}

# Upload video or html
blob?accept=video/*&accept=text/html

{
  "type": "permission",
  "resource": "blob",
  "accept": ["video/*", "text/html"]
}
```

## `account`

Control of PDS account hosting details, such as (private) account email.

Permissions of this type can not be included in permission sets, and must be requested directly by client apps.

Parameters:

> - `attr` (string, required): a component of account configuration. Wildcard is not supported.

Reviewer 2: ```suggestion
- `attr` (string, required, positional): a component of account configuration. Wildcard is not supported.
```

- `action` (string, optional): degree of control. Currently supports `read` or `manage`. If not specified, default is `read`.

Attributes:

- `email`: account email address. The `read` action makes the email and verification status visible. The `manage` action includes `read`, and also allows changing the email address.
- `repo`: ability to update entire public repository using a CAR file. The `manage` action allows importing entire CAR files, for example during account migration. The `read` action does nothing.

> - `status`: account status in network. The `manage` action allows deactivating and re-activating the account. The `read` action does nothing. (NOTE: this attribute is not part of PDS reference implementation as of November 2025)

Reviewer: IIRC we mentioned this would be performed through the account management interface and never through 3rd party apps/clients.

For example, the scope string `account:repo?action=manage` has a resource name `account`, a "positional" part with the value `repo`, and a key/value pair `action=manage`. Knowing that the positional parameter for `account` is `attr`, the same permission could be represented as `account?action=manage&attr=repo`.

Examples of syntactically and semantically valid scope strings:

```
identity:*
identity:*?
rpc?lxm=*&aud=did:web:api.example.com%23svc_appview
blob?accept=video/*&accept=text/html
repo:app.example.profile?action=create&action=update&action=delete
include:app.example.authFull?aud=did:web:api.example.com%23svc_chat
```

Scope strings that match the general syntax, but do not have valid semantics under the current resource definitions:

```
resource
resource:positional?key=val
resource:positional&thing?key=val
service:did:web:com.example#type?key=val
resource:
resource:?
resource:&
```

Examples which do not match the general syntax:

```
resource?
resource&
resource:positional?key=québec
emoji:☺️
```

## Permission Sets

> Full-featured client apps require a large number of granular permissions to function: dozens or even hundreds of individual permissions. This presents a user experience challenge, as long lists of permissions are difficult to parse, and a developer experience issue, as defining and maintaining these lists is toilsome.

Reviewer 2: yeah I think it's worth noting the security implications of avoiding bloated consent screens

This adds an intentional degree of temporal dynamism to auth sessions involving Permission sets. Lexicon designers can define new resources (eg, record types), update published sets to include permissions to those resources. Client software can then be updated to take advantage of those new lexicons, without requiring users to re-authenticate their sessions.

### Namespace Authority

Permission sets are limited to expressing permissions that reference resources under the same NSID namespace as the set itself.

Authority is based on the relative structure of NSIDs, without "siblings" or special namespaces. Broadly, this ensures that sets can not request permissions across namespaces. Specifically:

- Permission sets can address resources in the same NSID "group"; or "children" (sub-domains), recursively deep
- can not address "sibling groups" or "parents" in the NSID hierarchy

> For example, the set `app.example.feed.authOnlyPost` could include permissions to `app.example.feed.post` records and making `app.example.feed.getPostThread` API endpoint requests to remote services. But it could not grant permissions to `app.example.actor.profile`. A permission set `app.example.authFull`, which is a level up in the hierarchy, could include permissions to all these resources, or even further down the hierarchy.

Author: I added a paragraph about this, but in the section just above this

At the start of a session, if a permission set can not be resolved (and is not already in a local cache), the auth request will fail.

As an optional mechanism to trigger a cache purge, client instances can set an updated `software_version` field in their client metadata document. If the auth server sees a client version change, it should ensure that any permission sets referenced by the client would be refreshed. The motivation is to ensure that roll-out of new client software that requires updated sets can rely on auth servers giving the updated permissions within a timeframe shorter than the 24 hour "stale" cache lifetime.

## Usage and Implementation Guidelines

The currently defined resources and parameters do not make use of all of the syntax flexibility. Implementations might decide to focus on currently used syntax instead of the general syntax.

> Auth Servers may chose to use a "Lexicon Aggregator" service to offload lexicon resolution of permission sets. Such services would monitor a full-network firehose to detect updated permission sets. They could contain security logic to protect against NSID domain hijacking or adversarial changes to permission set scope.

Author: I think the second part of that is an OAuth semantic and should live in that spec (if it isn't already).

I'm not sure how to word the other part, or if we even really need to include it for now. In some sense the Auth Server can always do whatever it wants to protect the security of users.

I think I would lean towards not expanding on this in a very general way until we have something more concrete to describe or recommend.
